CodeMirror v6

Ryanseanlee opened the pull request:
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

Ryanseanlee committed:
Code mirror 6 checkpoint with AJV json schema validator

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	yarn.lock

Ryanseanlee committed:
yarn lock

RyanMG committed:
XH sass override for codeMirror editior needs it position to be respected.

 - update to code mirrors own CSS now uses !important for the same class
    with a relative position

Ryanseanlee committed:
AJV options

Ryanseanlee committed:
AJV options

RyanMG commented on desktop/cmp/input/CodeInput.scss:
```diff
@@ -46,40 +51,41 @@
-    & > .CodeMirror {
-      position: absolute;
+    & > .cm-editor {
+      position: absolute !important;
```
The latest CSS for code mirror added !important to some of their internal styling. For .cm-editor it made the position relative !important, breaking our needed override here.

RyanMG commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -4,15 +4,15 @@
-import 'codemirror/mode/javascript/javascript';
+import Ajv, {ValidateFunction} from 'ajv';
 import {codeInput, CodeInputProps} from './CodeInput';
-import {jsonlint} from './impl/jsonlint';
```
Was this the only use of jsonlint? Can be be removed?

RyanMG commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -4,15 +4,20 @@
 
-export type JsonInputProps = CodeInputProps;
+export interface JsonInputProps extends CodeInputProps {
+    jsonSchema?: PlainObject;
```
I did some testing in toolbox where I defined the schema and saw it working. It was not too hard, but it might be nice to add a comment here with some basics about the shape of the validation schema, and/or links to the AJV page showing how to get started with writing the validation.

Actually typing it seems like a nightmare, but we should provide some initial steps to get people started using it.

Ryanseanlee committed:
Type and comment jsonSchema in JsonInput.ts

Ryanseanlee committed:
Use codeMirror language-data to apply language extensions

Ryanseanlee committed:
remove console.log

Ryanseanlee committed:
update CHANGELOG

PeteDarinzo commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -240,210 +282,227 @@ class CodeInputModel extends HoistInputModel {
-    onChange = ev => {
-        this.noteValueChange(ev.target.value);
-    };
+    manageCodeEditor = async (container: HTMLElement) => {
```
Consider renaming to something like "createCodeEditor", since this method is doing the instantiation, and it seems clearer when providing it as a ref callback in the input cmp.

PeteDarinzo commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -240,210 +282,227 @@ class CodeInputModel extends HoistInputModel {
+        return XH.darkTheme ? dracula : solarizedLight;
+    }
+
+    private async getLanguageExtensionAsync(lang: string) {
```
Type this method's output: `Promise<LanguageSupport`. 

Returning `null` instead of `[]` in the unhappy path will fix type errors, and they will be filtered out the same on line 479.

PeteDarinzo commented on package.json:
```diff
@@ -115,7 +124,8 @@
     "@types/react": "18.x",
-    "@types/react-dom": "18.x"
+    "@types/react-dom": "18.x",
+    "@codemirror/view": "6.38.6"
```
Why is the resolution required?

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	yarn.lock

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

PeteDarinzo reviewed (commented):
Reviewed with comment

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

Ryanseanlee committed:
yarn.lock

Ryanseanlee commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -240,210 +282,227 @@ class CodeInputModel extends HoistInputModel {
     toggleFullScreen() {
         this.modalSupportModel.toggleIsModal();
-
-        // 'Nudge' the mouse wheel to trigger CodeMirror to update scrollbar state
```
Actually i just tested this and i think we can remove this 'nudge'  of the scrollbar. Not sure why it was needed before  but the scrollbar seems to be updated and working as expected

Ryanseanlee commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -471,11 +530,7 @@ const inputCmp = hoistCmp.factory<CodeInputModel>(({model, ...props}, ref) =>
-                    inputRef: model.manageCodeEditor,
-                    onChange: model.onChange
-                })
+                ref: model.manageCodeEditor
```
I just tried this and got weird rendering issues :/ 
<img width="286" height="685" alt="image" src="https://github.com/user-attachments/assets/a553df00-a7ea-4a1a-825f-3fe8e3774e49" />
looks like its rendering twice? or more?

Ryanseanlee committed:
Pete CR

amcclain commented:
> if its current use is limited it may be practical to add props to support whatever it was used for (toggling line numbers, providing keymaps).

Great, fully agree with that.  In general it's nice to have an "escape hatch" that allows us to set any underlying props that the library component might support - is that general concept not supportable with the upgraded version?  Doesn't need to be a blocker but would be interested to understand better

amcclain commented:
@PeteDarinzo / @Ryanseanlee - checking in on this upgrade - see my last question re. understanding why we can't continue to have some kind of pass-through "escape hatch" to the underlying codemirror.

Ryanseanlee committed:
Merge branch 'refs/heads/develop' into codemirror6&JsonSchema

# Conflicts:
#	yarn.lock

Ryanseanlee committed:
Add `lineNumbers` and lineWrapping prop

PeteDarinzo committed:
Limit code mirror themes to github light/dark

PeteDarinzo committed:
Yarn lock

PeteDarinzo committed:
Custom theme implementation

Ryanseanlee committed:
Statically import languages

Ryanseanlee committed:
test

Ryanseanlee committed:
use static imports add log for debug

Ryanseanlee committed:
debugging

Ryanseanlee committed:
debugging

Ryanseanlee committed:
debugging

Ryanseanlee committed:
debugging

Ryanseanlee committed:
remove custom theme impl

PeteDarinzo commented on CHANGELOG.md:
```diff
@@ -143,6 +143,32 @@
+* Hoist v77 **upgrades Code Mirror to v6** (from v5)
+  * editorProps deprecated:
+    The v5-style editorProps object (which accepted any CodeMirror config keys) is no longer
+    supported. Most former options can be achieved via v6 extensions or by setting CodeInput props
```
Remove reference to extensions since we are only supporting specific props

PeteDarinzo commented on desktop/cmp/input/CodeInput.scss:
```diff
@@ -46,40 +51,41 @@
-        box-sizing: content-box !important;
+      .cm-scroller {
+        overflow: auto;
+        //box-sizing: content-box !important;
```
Remove line 68 if cruft.

Ryanseanlee committed:
Pete CR clean ups

PeteDarinzo reviewed (approved):
Approved

lbwexler commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -4,15 +4,20 @@
+     * Configuration object with any properties supported by the AJV API.
+     * @see {@link https://ajv.js.org/options.html}
+     */
+    ajvProps?: Options;
```
can we call these `ajvSchema` `ajvOptions` ?

Would like to be consistent.  Also, don't think we need to introduce the react word 'props' here

lbwexler commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -101,6 +112,15 @@ export interface CodeInputProps extends HoistProps, HoistInputProps, LayoutProps
+    highlightActiveLine?: boolean;
+
+    /** True (default) to add line numbers to gutter. */
+    lineNumbers?: boolean | PlainObject;
```
No better typing available for the object form?  Can we comment on it?

lbwexler commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -214,21 +234,42 @@ class CodeInputModel extends HoistInputModel {
 
     override select() {
-        this.editor?.execCommand('selectAll');
+        if (!this.editor) return;
```
use the `?` syntax for consistency with implementations above

Ryanseanlee committed:
Merge Develop

Ryanseanlee commented on CHANGELOG.md:
```diff
@@ -21,6 +34,16 @@
     * `GroupingChooserProps.popoverTitle` - use `editorTitle`
     * `RelativeTimestampProps.options` - provide directly as top-level props
 
+### 📚 Libraries
```
Yes unfortunately I couldn't find another way due to the CM6 being fully modular by design.

Ryanseanlee committed:
Lee CR

lbwexler committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	CHANGELOG.md

Ryanseanlee committed:
Lee CR

Ryanseanlee committed:
Lee CR

amcclain commented on CHANGELOG.md:
```diff
@@ -2,6 +2,30 @@
+### 🎁 New Features
+* Added `jsonSchema` and `ajvProps` configs to JsonInputProps.
+    * `jsonSchema` - Used to validate the input Json
+    * `ajvProps` - Used to configure AJV Object
```
I admit I have no idea what AJV is ;)

amcclain commented on package.json:
```diff
@@ -42,9 +46,10 @@
     "@onsenui/fastclick": "~1.1.1",
     "@popperjs/core": "~2.11.0",
     "@seznam/compose-react-refs": "~1.0.5",
+    "@uiw/codemirror-theme-github": "^4.25.4",
```
Wrong semver spec - we've been allowing patch but not minor updates in most cases, as we've had situations where those introduce unexpected behaviors

Pls ensure you're clear on how these work

amcclain commented on desktop/cmp/filechooser/FileChooserModel.ts:
```diff
@@ -10,9 +10,8 @@ import {actionCol, calcActionColWidth} from '@xh/hoist/desktop/cmp/grid';
-import {isEmpty} from 'codemirror/src/util/misc';
 import {filesize} from 'filesize';
-import {find, uniqBy, without} from 'lodash';
+import {find, isEmpty, uniqBy, without} from 'lodash';
```
Good catch!!!

amcclain commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -214,21 +249,41 @@ class CodeInputModel extends HoistInputModel {
         super();
         makeObservable(this);
+
+        this.highlightStateFieldExtension = StateField.define<DecorationSet>({
```
I'd break this into its own method - it's enough clutter good to get out of the ctor - and then add a comment to explain what it's all about.

amcclain commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -214,21 +249,41 @@ class CodeInputModel extends HoistInputModel {
+            provide: f => EditorView.decorations.from(f)
+        });
+
         this.addReaction({
```
Minor thing - we can combine these `addReaction` calls - this was from before that took multiple specs

amcclain commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -276,173 +340,196 @@ class CodeInputModel extends HoistInputModel {
+                it => includes(it.alias, lang) || it.name.toLowerCase() === lang.toLowerCase()
+            );
+            if (!langDesc) return null;
+            return await langDesc.load();
```
What does this do again?  Have we tested it in a deployed scenario?  Does it async import a package from our own bundles (and does that work), or does it call out to someplace external to load code?

Ryanseanlee replied:
me and pete tested in a deployed scenario it works and it imports from our bundles and does not call an external server

amcclain commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -276,173 +340,196 @@ class CodeInputModel extends HoistInputModel {
+                languages,
+                it => includes(it.alias, lang) || it.name.toLowerCase() === lang.toLowerCase()
+            );
+            if (!langDesc) return null;
```
Would logError in this case as well vs. quietly returning null - dev has requested a language but used an unsupported identifier, is that right?

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	CHANGELOG.md

Ryanseanlee commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -214,21 +249,41 @@ class CodeInputModel extends HoistInputModel {
     }
 
     override focus() {
         this.editor?.focus();
```
Yeah unfortunately they dont have a blur on top-level editor. 
https://discuss.codemirror.net/t/code-mirror-6-has-focus-what-about-blur/4071

Ryanseanlee committed:
ATM CR

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	CHANGELOG.md
#	desktop/cmp/input/CodeInput.scss

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

# Conflicts:
#	CHANGELOG.md
#	package.json
#	yarn.lock

Ryanseanlee committed:
dry up

Ryanseanlee committed:
Merge branch 'develop' into codemirror6&JsonSchema

lbwexler committed:
Merge remote-tracking branch 'origin/develop' into codemirror6&JsonSchema

# Conflicts:
#	package.json
#	yarn.lock

lbwexler committed:
Remove AJV schema validation from JsonInput

Split AJV support out of the CodeMirror v6 migration to keep this PR
focused on the framework upgrade. JsonInput reverts to the prior
jsonlint-only linter (adapted for CM6 offset-based annotations).

AJV work preserved on `jsonInput-ajv-schema` branch for a follow-up PR.

lbwexler committed:
Align CodeMirror versions in CHANGELOG with package.json pins

lbwexler committed:
Restore case-insensitive search in CodeInput

CM5's getSearchCursor with caseFold=true was case-insensitive; the CM6
port via doc.indexOf lost that. Lowercase both sides to match the
documented behavior of enableSearch.

lbwexler committed:
Honor disabled at editor init in CodeInput

CM5 set readOnly from (disabled || readonly) at editor creation. The
CM6 port only considered readonly initially, relying on a reaction to
catch disabled. That left a one-tick window where a disabled editor
could be edited. Honor disabled in the initial extension config too.

lbwexler committed:
Bump CodeMirror pins to match resolved versions
